### PR TITLE
JAMES-2013 Extract Tika from James

### PR DESCRIPTION
**Pull request by chibenwa**
This PR is a rework of: https://github.com/linagora/james-project/pull/775

With https://github.com/linagora/james-project/pull/808 we can avoid depending on TIKA for anything related to JMAP, and thus avoid launching the tika container on any test involving JMAP, or a James server.

I still added a test to ensure James server is starting well with a tika container.

### CODE REVIEW

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    public ParsedContent extractContent(InputStream inputStream, String contentType) throws Exception {
+        ContentAndMetadata contentAndMetadata = convert(tikaHttpClient.rmetaAsJson(inputStream, contentType));
+        return new ParsedContent(contentAndMetadata.getContent(), contentAndMetadata.getMetadata());
+    }
+
+    private ContentAndMetadata convert(InputStream json) throws IOException, JsonParseException, JsonMappingException {
+        return objectMapper.readValue(json, ContentAndMetadata.class);
+    }
+
+    @VisibleForTesting static class ContentAndMetadataDeserializer extends JsonDeserializer<ContentAndMetadata> {
```
Comment (strokyl) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: I would add a break line between `@VisibleForTesting` and `static class`

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

File: mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java [hunk 2]
```diff
@@ -121,5 +151,76 @@ public void excelTest() throws Exception {
...
+                return new ByteArrayInputStream("[\"value1\"]".getBytes(Charsets.UTF_8));
+            }
+        });
+
+        InputStream inputStream = null;
+        textExtractor.extractContent(inputStream, "text/plain", "fake.txt");
+    }
+
+    @Test
+    public void asListOfStringShouldReturnASingletonWhenOneElement() {
```
Comment (quynhn) [inline mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java]: "Single" or "Singleton". What does it mean "Singleton" in this context?

Comment (chibenwa) [inline mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java]: "Singleton" means "single element"

Exemple from java API:

```
        Collections.singletonList(new Object())
```

Si IMO naming here is OK.

**Comment by chibenwa** [comment]
After an oral remark from @quynhn , we need to update documentation for Tika.

 - Appropriate location on the website, note on the README.

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaHttpClient.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.mailbox.tika;
+
+import java.io.InputStream;
+
+public interface TikaHttpClient {
+
+    InputStream rmetaAsJson(InputStream inputStream, String contentType) throws TikaException;
```
Comment (rouazana) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaHttpClient.java]: recursiveMetaData?

Comment (rouazana) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaHttpClient.java]: and same in implementation

**Comment by chibenwa** [comment]
Documentation added.

Screenshot: 
screenshot from 2017-06-07 11-22-37

Note please that I corrected the title on tika section...

File: src/site/xdoc/server/config-elasticsearch.xml [hunk 4]
```diff
@@ -87,6 +87,32 @@
...
+            <dd>IP or domain name of your Tika server. Defaults to 127.0.0.1</dd>
+            <dt><strong>tika.port</strong></dt>
+            <dd>Port of your tika server. Defaults to 9998</dd>
+            <dt><strong>tika.timeoutInMillis</strong></dt>
+            <dd>Timeout when issuing request to the tika server. Default is 3 seconds.</dd>
+        </dl>
+
+        Note: You can launch a tika server using this command line:
+
+        <code><pre>docker run --name tika logicalspark/docker-tikaserver:latest</pre></code>
```
Comment (rouazana) [inline src/site/xdoc/server/config-elasticsearch.xml]: Are we sure the api won't break in the future?

Comment (mbaechler) [inline src/site/xdoc/server/config-elasticsearch.xml]: we should not rely on "latest" anyway

Comment (chibenwa) [inline src/site/xdoc/server/config-elasticsearch.xml]: I can remove the docker line, but where the user gonna find which server to launch?

+ Seems docker is pretty hard for them, a quitte guided approach seems fine for me.

I will switch to **1.14** version here and in the tests.

Comment (chibenwa) [inline src/site/xdoc/server/config-elasticsearch.xml]: The tags were added 7 days ago <3

Comment (rouazana) [inline src/site/xdoc/server/config-elasticsearch.xml]: yes, the remark was just about that: documenting a fixed tag

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
```
[Building Java 8 james-2013-79d9ce9a-0764-4e28-b745-70ea014cfc93] Results :
[Building Java 8 james-2013-79d9ce9a-0764-4e28-b745-70ea014cfc93] 
[Building Java 8 james-2013-79d9ce9a-0764-4e28-b745-70ea014cfc93] Tests in error: 
[Building Java 8 james-2013-79d9ce9a-0764-4e28-b745-70ea014cfc93]   TikaTextExtractorTest.org.apache.james.mailbox.tika.TikaTextExtractorTest » ContainerLaunch
```

**Comment by chibenwa** [comment]
I try to launch the container manually to see why it fails

**Comment by chibenwa** [comment]
```
benwa@horizon ~/Documents/apache/onssd/james-project (git)-[JAMES-2013] % docker run --name tika logicalspark/docker-tikaserver:1.14  
Unable to find image 'logicalspark/docker-tikaserver:1.14' locally
1.14: Pulling from logicalspark/docker-tikaserver
b6f892c0043b: Pull complete 
55010f332b04: Pull complete 
2955fb827c94: Pull complete 
3deef3fcbd30: Pull complete 
cf9722e506aa: Pull complete 
18fe3e1358cd: Pull complete 
Digest: sha256:279af8b3cd7f747d2fb287b787a42affc8e37855ced7cedf701963d1f761f94a
Status: Downloaded newer image for logicalspark/docker-tikaserver:1.14
Error: Invalid or corrupt jarfile /tika-server-1.14.jar
```

**Comment by chibenwa** [comment]
Using tag **1.15rc1** works

**Comment by chibenwa** [comment]
With this version:

```
[Building Java 8 james-2013-a08216ee-26c2-4fc8-a472-c87b47e20188] Tests in error: 
[Building Java 8 james-2013-a08216ee-26c2-4fc8-a472-c87b47e20188]   TikaTextExtractorTest.documentWithBadDeclaredMetadataShouldBeWellHandled:97 » IllegalState
[Building Java 8 james-2013-a08216ee-26c2-4fc8-a472-c87b47e20188]   TikaTextExtractorTest.odsTest:129 » IllegalState The response should have only...
[Building Java 8 james-2013-a08216ee-26c2-4fc8-a472-c87b47e20188]   TikaTextExtractorTest.slideOdpTest:113 » IllegalState The response should have...
[Building Java 8 james-2013-a08216ee-26c2-4fc8-a472-c87b47e20188]   TikaTextExtractorTest.textOdtTest:89 » IllegalState The response should have o..
```

**Comment by chibenwa** [comment]
Some documents might generate multiple parsing answers. The main one is the first one.

EG:

```
[{
	"Content-Type": "application/vnd.oasis.opendocument.spreadsheet",
	"Creation-Date": "2015-03-09T16:35:25.926263921",
	"Edit-Time": "P0D",
	"Last-Modified": "2015-03-09T16:36:05.298706046",
	"Last-Save-Date": "2015-03-09T16:36:05.298706046",
	"Object-Count": "0",
	"Table-Count": "1",
	"X-Parsed-By": ["org.apache.tika.parser.DefaultParser", "org.apache.tika.parser.odf.OpenDocumentParser"],
	"X-TIKA:content": "\n\n\n\n\n\n\n\tThis is an aesome LibreOffice document !\n\n\n???\nPage \n??? (???)\n00/00/0000, 00:00:00\nPage  / \n",
	"X-TIKA:parse_time_millis": "217",
	"date": "2015-03-09T16:36:05.298706046",
	"dcterms:created": "2015-03-09T16:35:25.926263921",
	"dcterms:modified": "2015-03-09T16:36:05.298706046",
	"editing-cycles": "1",
	"generator": "LibreOffice/4.2.7.2$Linux_X86_64 LibreOffice_project/420m0$Build-2",
	"meta:creation-date": "2015-03-09T16:35:25.926263921",
	"meta:object-count": "0",
	"meta:save-date": "2015-03-09T16:36:05.298706046",
	"meta:table-count": "1",
	"modified": "2015-03-09T16:36:05.298706046",
	"nbObject": "0",
	"nbTab": "1"
}, {
	"Chroma BlackIsZero": "true",
	"Chroma ColorSpaceType": "RGB",
	"Chroma NumChannels": "3",
	"Compression CompressionTypeName": "deflate",
	"Compression Lossless": "true",
	"Compression NumProgressiveScans": "1",
	"Content-Type": "image/png",
	"Data BitsPerSample": "8 8 8",
	"Data PlanarConfiguration": "PixelInterleaved",
	"Data SampleFormat": "UnsignedIntegral",
	"Dimension ImageOrientation": "Normal",
	"Dimension PixelAspectRatio": "1.0",
	"IHDR": "width=189, height=255, bitDepth=8, colorType=RGB, compressionMethod=deflate, filterMethod=adaptive, interlaceMethod=none",
	"Transparency Alpha": "none",
	"X-Parsed-By": ["org.apache.tika.parser.DefaultParser", "org.apache.tika.parser.ocr.TesseractOCRParser", "org.apache.tika.parser.image.ImageParser"],
	"X-TIKA:embedded_resource_path": "/embedded-1",
	"X-TIKA:origResourceName": "Thumbnails/thumbnail.png",
	"X-TIKA:parse_time_millis": "200",
	"height": "255",
	"tiff:BitsPerSample": "8 8 8",
	"tiff:ImageLength": "255",
	"tiff:ImageWidth": "189",
	"width": "189"
}]
```

We should then always return the first one, and ignore following answers.

File: src/site/xdoc/server/config-elasticsearch.xml [hunk 5]
```diff
@@ -87,6 +87,32 @@
...
+
+        <p>Note that to use this feature you need Guice, built with ElasticSearch</p>
+
+        <p>Consult <a href="https://github.com/apache/james-project/blob/master/dockerfiles/run/guice/cassandra/destination/conf/tika.properties">tika.properties</a> to get some examples and hints.</p>
+
+        Here are the different properties:
+
+        <dl>
+            <dt><strong>tika.host</strong></dt>
+            <dd>IP or domain name of your Tika server. Defaults to 127.0.0.1</dd>
```
Comment (strokyl) [inline src/site/xdoc/server/config-elasticsearch.xml]: "Default is 127.0.0.1"

File: src/site/xdoc/server/config-elasticsearch.xml [hunk 6]
```diff
@@ -87,6 +87,32 @@
...
+            <dt><strong>tika.port</strong></dt>
+            <dd>Port of your tika server. Defaults to 9998</dd>
```
Comment (strokyl) [inline src/site/xdoc/server/config-elasticsearch.xml]: "Default is 9998"

File: src/site/xdoc/server/config-elasticsearch.xml [hunk 7]
```diff
@@ -87,6 +87,32 @@
...
+            <dt><strong>tika.timeoutInMillis</strong></dt>
+            <dd>Timeout when issuing request to the tika server. Default is 3 seconds.</dd>
```
Comment (strokyl) [inline src/site/xdoc/server/config-elasticsearch.xml]: "Default value is 3 seconds"

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**